lib/Sema/TypeCheckProtocolInference.cpp

```diff
@@ -1200,23 +1202,26 @@ AssociatedTypeDecl *AssociatedTypeInference::completeSolution(
         if (type->hasError())
           return witness.getAssocType();
 
-        // FIXME: If we still have a type parameter and it isn't a generic
-        // parameter of the conforming nominal, it's either a cycle or a
-        // solution that is beyond the current algorithm, i.e.
-        //
+        // FIXME: If mapping into context yields an error, or we still have a
+        // type parameter despite not having a generic environment, then a type
+        // parameter was sent to a tentative type witness that itself is a type
+        // parameter, and the solution is cyclic, e.g. { A := B.A, B := A.B },
+        // or beyond the current algorithm, e.g.
         // protocol P {
         //   associatedtype A = B
         //   associatedtype B = C
         //   associatedtype C = Int
         // }
         // struct Conformer: P {}
-        if (type->hasTypeParameter() &&
-            !adoptee->getAnyNominal()->isGeneric()) {
+        if (dc->getGenericEnvironmentOfContext()) {
```

Comment:
I think isGenericContext() is cheaper since you're just checking if the result here is null, but it probably doesn't matter since there's a mapTypeIntoContext() call immediately following.

Reply:
I think the difference is negligible for the reason you stated. Can we change it alongside the upcoming improvements to this part of associated type inference that we discussed recently? Had to refactor this method anyway.

Reply:
Yeah of course.